File: src/agentsync/cli/install_hook.py
"""Install a git pre-commit hook that runs `agentsync sync --check`."""

from __future__ import annotations

import stat
from pathlib import Path
from typing import Annotated

import typer
from rich.console import Console

from agentsync.cli._exit import ExitCode
from agentsync.core.paths import resolve_paths

console = Console()

HOOK_MARKER = "# agentsync:pre-commit"

HOOK_SCRIPT = f"""#!/usr/bin/env sh
{HOOK_MARKER}
# Block commits whose generated agent files have drifted from /agent.
# Edit /agent/* and run `agentsync sync`, or remove this hook to bypass.

if ! command -v agentsync >/dev/null 2>&1; then
    echo "agentsync not on PATH — skipping check"
    exit 0
fi

agentsync sync --check
"""


def install_hook_command(
    repo_root: Annotated[
        Path | None,
        typer.Option("--repo", help="Repo root (defaults to cwd).", file_okay=False),
    ] = None,
    force: Annotated[
        bool,
        typer.Option("--force", "-f", help="Overwrite an existing pre-commit hook."),
    ] = False,
) -> None:
    paths = resolve_paths(repo_root)
    git_dir = paths.repo_root / ".git"
    if not git_dir.is_dir():
        console.print(
            "[red]error:[/red] not a git repository. Run `git init` first or use "
            "[bold]pre-commit[/bold] (M4 ships an official hook)."
        )
        raise typer.Exit(ExitCode.USER_ERROR)

    hooks_dir = git_dir / "hooks"
    hooks_dir.mkdir(parents=True, exist_ok=True)
    hook_file = hooks_dir / "pre-commit"

    if hook_file.exists() and not force:
        existing = hook_file.read_text(encoding="utf-8", errors="replace")
        if HOOK_MARKER in existing:
            console.print(f"[yellow]ok[/yellow] hook already installed at {hook_file}")
            raise typer.Exit(ExitCode.OK)
        console.print(
            f"[red]error:[/red] {hook_file} already exists. Re-run with --force to overwrite."
        )
        raise typer.Exit(ExitCode.USER_ERROR)

    hook_file.write_text(HOOK_SCRIPT, encoding="utf-8")
    mode = hook_file.stat().st_mode
    hook_file.chmod(mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
    rel = hook_file.relative_to(paths.repo_root) if hook_file.is_relative_to(paths.repo_root) else hook_file
    console.print(f"[green]installed[/green] {rel}")
    console.print("  → runs [bold]agentsync sync --check[/bold] before every commit.")

File: src/cli/installHook.ts
import path from "node:path";
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile, chmod } from "node:fs/promises";
import { type Logger } from "../core/logger.js";
import { UserError } from "../core/errors.js";
import { resolveCwd } from "../core/paths.js";
import { isGitRepo } from "../core/git.js";

export type HookType = "native" | "husky";

export interface InstallHookOptions {
  json?: boolean;
}

const NATIVE_HOOK_SCRIPT = `#!/bin/sh
# Installed by agentsync install-hook
agentsync sync --check
`;

const HUSKY_HOOK_SCRIPT = `agentsync sync --check
`;

export async function installHookCommand(
  cwd: string | undefined,
  hookType: HookType,
  opts: InstallHookOptions,
  logger: Logger,
): Promise<void> {
  const root = resolveCwd(cwd);

  if (!(await isGitRepo(root))) {
    throw new UserError(`Not a git repository: ${root}. Initialize git first.`);
  }

  if (hookType === "husky") {
    await installHuskyHook(root, opts, logger);
  } else {
    await installNativeHook(root, opts, logger);
  }
}

async function installNativeHook(
  root: string,
  opts: InstallHookOptions,
  logger: Logger,
): Promise<void> {
  const hooksDir = path.join(root, ".git", "hooks");
  const hookPath = path.join(hooksDir, "pre-commit");

  await mkdir(hooksDir, { recursive: true });

  if (existsSync(hookPath)) {
    const existing = await readFile(hookPath, "utf8");
    if (existing.includes("agentsync sync --check")) {
      if (opts.json) {
        logger.print(JSON.stringify({ ok: true, hookPath, action: "already-installed" }, null, 2));
      } else {
        logger.info(`Hook already installed at ${hookPath}`);
      }
      return;
    }
    // Append to existing hook
    const updated = existing.trimEnd() + "\n\n" + NATIVE_HOOK_SCRIPT;
    await writeFile(hookPath, updated, "utf8");
  } else {
    await writeFile(hookPath, NATIVE_HOOK_SCRIPT, "utf8");
  }

  await chmod(hookPath, 0o755);

  if (opts.json) {
    logger.print(JSON.stringify({ ok: true, hookPath, action: "installed" }, null, 2));
  } else {
    logger.success(`Installed native pre-commit hook at ${hookPath}`);
  }
}

async function installHuskyHook(
  root: string,
  opts: InstallHookOptions,
  logger: Logger,
): Promise<void> {
  const huskyDir = path.join(root, ".husky");
  const hookPath = path.join(huskyDir, "pre-commit");

  if (!existsSync(huskyDir)) {
    throw new UserError(
      `Husky is not initialized. Run \`npx husky init\` first, then re-run \`agentsync install-hook husky\`.`,
    );
  }

  if (existsSync(hookPath)) {
    const existing = await readFile(hookPath, "utf8");
    if (existing.includes("agentsync sync --check")) {
      if (opts.json) {
        logger.print(JSON.stringify({ ok: true, hookPath, action: "already-installed" }, null, 2));
      } else {
        logger.info(`Hook already installed at ${hookPath}`);
      }
      return;
    }
    const updated = existing.trimEnd() + "\n" + HUSKY_HOOK_SCRIPT;
    await writeFile(hookPath, updated, "utf8");
  } else {
    await writeFile(hookPath, `#!/bin/sh\n${HUSKY_HOOK_SCRIPT}`, "utf8");
    await chmod(hookPath, 0o755);
  }

  if (opts.json) {
    logger.print(JSON.stringify({ ok: true, hookPath, action: "installed" }, null, 2));
  } else {
    logger.success(`Installed Husky pre-commit hook at ${hookPath}`);
    logger.info(`Make sure \`agentsync\` is in your PATH or use \`npx agentsync sync --check\`.`);
  }
}
